README.md

## The `esaps` package <img src="man/figures/logo.png" style="margin-left:10px;margin-bottom:5px;" width="160" align="right"></a>

<!-- badges: start -->

[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version-ago/esaps)](https://cran.r-project.org/package=esaps)
[![](https://img.shields.io/badge/devel%20version-0.2.2-blue.svg)](https://github.com/Nicolas-Schmidt/esaps)
[![R build
status](https://github.com/Nicolas-Schmidt/esaps/workflows/R-CMD-check/badge.svg)](https://github.com/Nicolas-Schmidt/esaps/actions)
[![Project Status: Active – The project has reached a stable, usable
state and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN RStudio
mirrordownloads](https://cranlogs.r-pkg.org/badges/grand-total/esaps?color=blue)](https://www.r-pkg.org/pkg/esaps)

<!-- badges: end -->

### Description

It allows structuring electoral data of different size and structure to
calculate various indicators frequently used in the studies of electoral
systems and party systems. Indicators of electoral volatility, electoral
disproportionality, party nationalization and the effective number of
parties are included.

### Installation

``` r
# Install speech from CRAN
install.packages("esaps")

# The development version from GitHub:
if (!require("remotes")) install.packages("remotes")
remotes::install_github("Nicolas-Schmidt/esaps")
```

### Functions

| Indicator                    | Method                                                | Variables                                     | Function                    |
|------------------------------|-------------------------------------------------------|-----------------------------------------------|-----------------------------|
| Electoral Volatility         | [Pedersen (1979)](http://bit.ly/2p4cyZQ)              | `election` `unit` `party` `votes`             | `evolat(., method=1)`       |
|                              | [Powell and Tucker (2014)](http://bit.ly/2FpoeBm)     | `election` `unit` `party` `votes`             | `evolat(., method=2)`       |
|                              | [Torcal and Lago (2015)](http://bit.ly/2FD83zE)       | `election` `unit` `party` `votes` `M`         | `evolat(., method=3)`       |
| Effective Number of Parties  | [Laakso and Taagepera (1979)](http://bit.ly/2oYU4dV)  | `election` `unit` `party` `votes` or `seats`  | `enp()`                     |
| Party System Nationalization | [Mainwaring and Jones (2003)](https://bit.ly/2w9QgtO) | `election` `unit` `party` `votes` `votes_nac` | `psn(., method=1)`          |
|                              | [Chhibber and Kollman (2004)](https://bit.ly/2oj76SY) | `election` `unit` `party` `votes` `votes_nac` | `psn(., method=2)`          |
| Party Nationalization        | [Mainwaring and Jones (2003)](https://bit.ly/2w9QgtO) | `election` `unit` `party` `votes`             | `psn(., method=1, pn=TRUE)` |
| Electoral Disproportionality | Rae (1971)                                            | `election` `unit` `party` `votes` `seats`     | `dispro(., method=1)`       |
|                              | [Loosemore and Hanby (1971)](https://bit.ly/2B1mIDH)  | `election` `unit` `party` `votes` `seats`     | `dispro(., method=2)`       |
|                              | Lijphart\_1 (1986)                                    | `election` `unit` `party` `votes` `seats`     | `dispro(., method=3)`       |
|                              | [Lijphart\_2 (1994)](https://bit.ly/2McLshB)          | `election` `unit` `party` `votes` `seats`     | `dispro(., method=4)`       |
|                              | [Gallagher (1991)](https://bit.ly/2M6CPoy)            | `election` `unit` `party` `votes` `seats`     | `dispro(., method=5)`       |
|                              | [Cox and Shugart (1991)](https://bit.ly/2M9Bttr)      | `election` `unit` `party` `votes` `seats`     | `dispro(., method=6)`       |

### Indicators and Variables

<img src="man/figures/net.png" style="margin-center:30px;margin-bottom:5px;" width="1160" align="center"></a>

### Examples

#### Load data

``` r
library(esaps)

votes <- list(data.frame(country = rep("ARG", 3),
                         year = c(1995, 2000, 2005),
                         party_A = c(40,10,20),
                         party_B = c(35,20,40),
                         party_C = c(25,70,40)),
              data.frame(country = rep("URY", 4),
                         year = c(1995, 2000, 2005, 2010),
                         party_A = c(30,30,20,20),
                         party_B = c(30,50,40, 30),
                         party_C = c(30,10,30, 25),
                         party_D = c(10,10,10,25)),
              data.frame(country = rep("BRA", 2),
                         year = c(1993, 1998),
                         party_A = c(30, 55),
                         party_B = c(70, 45)))

votes
#> [[1]]
#>   country year party_A party_B party_C
#> 1     ARG 1995      40      35      25
#> 2     ARG 2000      10      20      70
#> 3     ARG 2005      20      40      40
#> 
#> [[2]]
#>   country year party_A party_B party_C party_D
#> 1     URY 1995      30      30      30      10
#> 2     URY 2000      30      50      10      10
#> 3     URY 2005      20      40      30      10
#> 4     URY 2010      20      30      25      25
#> 
#> [[3]]
#>   country year party_A party_B
#> 1     BRA 1993      30      70
#> 2     BRA 1998      55      45

votes <- convert_esaps(dataset = votes, unit.name = "country", election.name = "year")

votes
#>    election unit   party votes
#> 1      1995  ARG party_A    40
#> 2      2000  ARG party_A    10
#> 3      2005  ARG party_A    20
#> 4      1995  ARG party_B    35
#> 5      2000  ARG party_B    20
#> 6      2005  ARG party_B    40
#> 7      1995  ARG party_C    25
#> 8      2000  ARG party_C    70
#> 9      2005  ARG party_C    40
#> 10     1993  BRA party_A    30
#> 11     1998  BRA party_A    55
#> 12     1993  BRA party_B    70
#> 13     1998  BRA party_B    45
#> 14     1995  URY party_A    30
#> 15     2000  URY party_A    30
#> 16     2005  URY party_A    20
#> 17     2010  URY party_A    20
#> 18     1995  URY party_B    30
#> 19     2000  URY party_B    50
#> 20     2005  URY party_B    40
#> 21     2010  URY party_B    30
#> 22     1995  URY party_C    30
#> 23     2000  URY party_C    10
#> 24     2005  URY party_C    30
#> 25     2010  URY party_C    25
#> 26     1995  URY party_D    10
#> 27     2000  URY party_D    10
#> 28     2005  URY party_D    10
#> 29     2010  URY party_D    25

## VOTES AND SEATS

votes_2 <- data.frame(
        year = c(2000, 2005), 
        country = "URY", 
        votes_party1 = c(20, 30), 
        votes_party2 = c(30, 35),
        votes_party3 = c(40, 25),
        votes_party4 = c(10, 10), 
        seats_party1 = c(25, 35),
        seats_party2 = c(20, 30),
        seats_party3 = c(40, 30),   
        seats_party4 = c(15, 5)   
)

votes_2
#>   year country votes_party1 votes_party2 votes_party3 votes_party4 seats_party1
#> 1 2000     URY           20           30           40           10           25
#> 2 2005     URY           30           35           25           10           35
#>   seats_party2 seats_party3 seats_party4
#> 1           20           40           15
#> 2           30           30            5

votes_2c <- convert_esaps(dataset = votes_2, unit.name = "country", election.name = "year", seats = TRUE)

votes_2c
#>   election unit        party votes seats
#> 1     2000  URY votes_party1    20    25
#> 2     2000  URY votes_party2    30    20
#> 3     2000  URY votes_party3    40    40
#> 4     2000  URY votes_party4    10    15
#> 5     2005  URY votes_party1    30    35
#> 6     2005  URY votes_party2    35    30
#> 7     2005  URY votes_party3    25    30
#> 8     2005  URY votes_party4    10     5
```

#### Functions

Electoral volatility

``` r
evolat(votes, 1)
#>   election unit eVolat
#> 1     2000  ARG     45
#> 2     2005  ARG     30
#> 3     1998  BRA     25
#> 4     2000  URY     20
#> 5     2005  URY     20
#> 6     2010  URY     15

evolat(tidy_data = votes, method = 1, summary = TRUE)
#> [[1]]
#>   election unit eVolat
#> 1     2000  ARG     45
#> 2     2005  ARG     30
#> 3     1998  BRA     25
#> 4     2000  URY     20
#> 5     2005  URY     20
#> 6     2010  URY     15
#> 
#> [[2]]
#>   unit first_elec last_elec election mean_volat sd_volat
#> 1  ARG       2000      2005        2      37.50    10.61
#> 2  BRA       1998      1998        1      25.00       NA
#> 3  URY       2000      2010        3      18.33     2.89
```

Effective Number of Parties

``` r
enp(votes)
#>   election unit  enp
#> 1     1995  ARG 2.90
#> 2     2000  ARG 1.85
#> 3     2005  ARG 2.78
#> 4     1993  BRA 1.72
#> 5     1998  BRA 1.98
#> 6     1995  URY 3.57
#> 7     2000  URY 2.78
#> 8     2005  URY 3.33
#> 9     2010  URY 3.92
```

Electoral Disproportionality

``` r
votes_2
#>   year country votes_party1 votes_party2 votes_party3 votes_party4 seats_party1
#> 1 2000     URY           20           30           40           10           25
#> 2 2005     URY           30           35           25           10           35
#>   seats_party2 seats_party3 seats_party4
#> 1           20           40           15
#> 2           30           30            5

v2 <- convert_esaps(dataset = votes_2, unit.name = "country", election.name = "year", seats = TRUE)
dispro(v2, 1:6, 1)
#>   election unit  Rae  LH Lijphart_1 Lijphart_2 Gallagher Cox_Shugart
#> 1     2000  URY 0.05 0.1       0.10       0.06      0.32        1.00
#> 2     2005  URY 0.05 0.1       0.05       0.06      0.32        0.73

library(magrittr)

convert_esaps(dataset = votes_2, unit.name = "country", election.name = "year", seats = TRUE) %>%
        dispro(1:3)
#>   election unit  Rae  LH Lijphart_1
#> 1     2000  URY 0.05 0.1       0.10
#> 2     2005  URY 0.05 0.1       0.05


dispro(tidy_data = v2, method = c("Rae", "Gallagher"))
#>   election unit  Rae Gallagher
#> 1     2000  URY 0.05      0.32
#> 2     2005  URY 0.05      0.32
```

Party System Nationalization Score and Party Nationalization Score

``` r
votes3 <- data.frame(election = rep(2000,4),
                    unit  = rep(c("District_1", "District_2"), each = 2),
                    party = rep(c("party_A", "party_B"), 2),
                    votes = c(0.60,0.40, 0.30, 0.70),
                    votes_nac = rep(c(0.55,0.45),2)
                    )

votes3
#>   election       unit   party votes votes_nac
#> 1     2000 District_1 party_A   0.6      0.55
#> 2     2000 District_1 party_B   0.4      0.45
#> 3     2000 District_2 party_A   0.3      0.55
#> 4     2000 District_2 party_B   0.7      0.45

psns(tidy_data = votes3, method = 1)
#>   election   psns
#> 1     2000 84.695

psns(tidy_data = votes3, method = 1, pns = TRUE)
#> $PSNS
#>   election   psns
#> 1     2000 84.695
#> 
#> $PNS
#>     party election   pns
#> 1 party_A     2000 0.833
#> 2 party_B     2000 0.864
```

Electoral volatility in uruguay

``` r
library(Boreluy) # https://nicolas-schmidt.github.io/Boreluy/
library(tidyverse)

map_df(c(1984, 1989, 1994, 1999, 2004), nacional_uy) %>% 
    as_esaps() %>% 
    evolat("Pedersen") %>% 
        ggplot(aes(x = election, y = eVolat)) +
        geom_line(aes(group = 1)) + 
        geom_point(size = 2, color = "red") +
        theme_minimal()
```

<img src="man/figures/README-unnamed-chunk-8-1.png" width="80%" />

Linear Model

``` r
library(tidyverse)

dat <- inner_join(evolat(votes, 1), enp(votes))

dat
#>   election unit eVolat  enp
#> 1     2000  ARG     45 1.85
#> 2     2005  ARG     30 2.78
#> 3     1998  BRA     25 1.98
#> 4     2000  URY     20 2.78
#> 5     2005  URY     20 3.33
#> 6     2010  URY     15 3.92

lm(formula(dat[,4:3]), data = dat)
#> 
#> Call:
#> lm(formula = formula(dat[, 4:3]), data = dat)
#> 
#> Coefficients:
#> (Intercept)       eVolat  
#>     4.25394     -0.05731

ggplot(dat, aes(x = enp, y = eVolat )) +
    geom_point() + 
    geom_smooth(method = lm, se = TRUE) +
    theme_minimal()
```

<img src="man/figures/README-f1-1.png" width="80%" />

### Citation

To cite `esaps` in publications, please use:

``` r
citation("esaps")
```

### Author

Nicolas Schmidt (<nschmidt@cienciassociales.edu.uy>)


NEWS.md
# esaps 0.2.2

* add `psns()`, `dispro()` and `convert_esaps()`

# esaps 0.1.0

* CRAN-RELEASE


R/enp.R
#' @title Effective Number of Parties
#'
#' @description The Effective Number of Parties (ENP) is an index developed by
#'     Laakso and Taagepera (1979) that allows to count the relevant parties in a
#'     party system. The formula consists on dividing one over the sum of the squares
#'     of the proportions (votes or seats) that the parties obtain in an electoral instance.
#' @param tidy_data data.frame that contains the following variables with these names:
#'    \itemize{
#'             \item{\code{election}: year of election.}
#'             \item{\code{unit}: the unit of analysis (province, department ...)}
#'             \item{\code{party}: name of the political parties that obtain votes.}
#'             \item{\code{votes}: votes obtained by each party.}
#'             \item{\code{seats}: .}
#'            }
#'    If the data is not structured in this way you can order it with: \code{\link{convert_esaps}}.
#' @param enp_seats enp_seats = TRUE allows us to calculate jointly the effective
#'     number of electoral parties and in the congress.
#' @param summary Summary of the data by unit, by default it is \code{FALSE}.
#'
#'
#' @return if \code{summary = FALSE,} return data.frame.
#'
#'     if \code{summary = TRUE}, return a list with two data.frame.
#'     \itemize{
#'             \item{\code{list[[1]]}} {Indicator}
#'             \item{\code{list[[2]]}} {Summary}
#'                   \itemize{
#'                       \item{\code{min} variable 'election'}
#'                       \item{\code{max} variable 'election'}
#'                       \item{number of elections}
#'                       \item{\code{mean} indicator}
#'                       }
#'      }
#'
#' @author Nicolas Schmidt \email{nschmidt@cienciassociales.edu.uy}
#'
#' @examples
#' votes <- data.frame(election = rep(c(2000, 2005), each = 4),
#'                    unit  = rep(c("ARG", "URY"), each = 4),
#'                    party = c("party_A", "party_B","party_C","party_D"),
#'                    votes = c(20, 20, 50, 10, 30, 35, 25, 10),
#'                    seats = c(25, 25, 40, 10, 30, 30, 30, 10)
#'                    )
#'
#' enp(votes)
#' enp(votes, enp_seats = TRUE)
#' enp(votes, summary = TRUE)
#'
#' @export


enp <- function(tidy_data,
                enp_seats = FALSE,
                summary = FALSE){

        if(!is.data.frame(tidy_data)){stop("'tidy_data' must be a 'data.frame'.", call. = FALSE)}

        vars <- c("election", "unit", "party", "votes", "seats")
        if(isTRUE(enp_seats)){
                if(all(vars %in% names(tidy_data))){
                        loc <- which(names(tidy_data) %in% vars[4:5])
                        colnames(tidy_data)[loc] <- c("enp", "enp_c")
                        vars[4:5] <- c("enp", "enp_c")
                        if(ncol(tidy_data) > 5){
                                tidy_data <- tidy_data[, vars]
                        }else{
                                tidy_data <- tidy_data[, vars]
                        }
                }else{
                        stop("The names of the variables should be: 'election', 'unit', 'party', 'votes' and 'seats'.", call. = FALSE)
                }
        }else{
                vars <- vars[-5]
                if(all(vars %in% names(tidy_data))){
                        loc <- which(names(tidy_data)%in%vars[4])
                        colnames(tidy_data)[loc] <- "enp"
                        vars[4] <- "enp"
                        if(ncol(tidy_data) > 4){
                                tidy_data <- tidy_data[, vars]
                        }else{
                                tidy_data <- tidy_data[, vars]
                        }
                }else{
                        stop("The names of the variables should be: 'election', 'unit', 'party' and 'votes'.", call. = FALSE)
                }
        }
        if(sum(is.na(tidy_data[, -3])) != 0){stop("The variable 'election','unit','votes' or 'seats' must not have NA values.", call. = FALSE)}
        v1     <- unlist(lapply(split(tidy_data, tidy_data$unit), function(x){split(x, x$election)}), recursive = FALSE)
        out    <- lapply(v1, function(x){round(apply(x[-c(1:3)], 2, ENP),2)})
        output <- cbind(do.call(rbind, lapply(v1,"[",1, 1:2)), do.call(rbind, out))
        output <- output[order(output$unit),]
        rownames(output) <- NULL

        if (summary==TRUE){
                if(isTRUE(enp_seats)){
                        tab <- plyr::ddply(output, ~ unit,
                                   function(x) c(min.election = min(x$election),
                                                 max.election = max(x$election),
                                                 n.election   = length(x$election),
                                                 mean_enp     = round(mean(x$enp), 2),
                                                 sd_enp       = round(stats::sd(x$enp), 2),
                                                 mean_enp_c   = round(mean(x$enp_c), 2),
                                                 sd_enp_c     = round(stats::sd(x$enp_c), 2)))
                        out_enp <- list(output, tab)
                        return(out_enp)
                }else{
                        tab <- plyr::ddply(output, ~ unit,
                                   function(x) c(min.election = min(x$election),
                                                 max.election = max(x$election),
                                                 n.election   = length(x$election),
                                                 mean_enp     = round(mean(x$enp), 2),
                                                 sd_enp       = round(stats::sd(x$enp), 2)))
                        out_enp <- list(output, tab)
                        return(out_enp)
                }
        }
        output
}












R/psns.R
#' @title Party System Nationalization Score
#'
#' @description Party System Nationalization Score Mainwaring and Jones (2003) and Chhibber and Kollman (2004)
#'
#' @param tidy_data data.frame that contains the following variables with these names:
#'    \itemize{
#'             \item{\code{election}: year of election.}
#'             \item{\code{unit}: the unit of analysis (province, department ...)}
#'             \item{\code{party}: name of the political parties that obtain votes.}
#'             \item{\code{votes}: votes obtained by each party.}
#'             \item{\code{votes_nac}: votes at national level for each party.}
#'            }
#'    If the data is not structured in this way you can order it with: \code{\link{convert_esaps}}.
#'
#' @param method Method to calculate Party System Nationalization Score:
#'     \itemize{
#'             \item{\code{method = "Mainwaring and Jones"} or \code{method = 1}.}
#'             \item{\code{method = "Chhibber and Kollman"} or \code{method = 2}.}
#'             }
#'
#' @param pns by default it is \code{FALSE}. If \code{TRUE}, the Party Nationalization Score
#'     is calculated. In method, you must indicate: \code{method = 1}.
#'
#' @param scale By default it is \code{100}, the indices will vary between 0 and 100.
#'     If \code{scale = 1} the variation will be between 0 and 1.
#'
#'
#'
#' @return if \code{pns = FALSE,} return data.frame.
#'
#'     if \code{pns = TRUE}, return a list with two data.frame.
#'     \itemize{
#'             \item{\code{list[[1]]}} {PSNS: Party System Nationalization Score}
#'             \item{\code{list[[2]]}} {PNS: Party Nationalization Score}
#'            }
#'
#' @author Nicolas Schmidt \email{nschmidt@cienciassociales.edu.uy}
#'
#'
#'
#' @examples
#' votes <- data.frame(election = rep(2000,4),
#'                     unit  = rep(c("District_1", "District_2"), each = 2),
#'                     party = rep(c("party_A", "party_B"), 2),
#'                     votes = c(0.60,0.40, 0.30, 0.70),
#'                     votes_nac = rep(c(0.55,0.45),2)
#'                    )
#' psns(tidy_data = votes, method = 1)
#' psns(tidy_data = votes, method = 1, pns = TRUE)
#'
#' @export
#'



psns <- function(tidy_data,
                method,
                pns = FALSE,
                scale = 100){

        tidy_data <- tidy_data[, names(tidy_data) %in% c('election','unit', 'party', 'votes', 'votes_nac')]
        tidy_data <- stats::na.omit(tidy_data)
        if(!is.data.frame(tidy_data)){stop("'tidy_data' must be a 'data.frame'.", call. = FALSE)}
        if(missing(method)){stop("You must select only one method.", call. = FALSE)}
        if(sum(is.na(tidy_data[, 1:2])) != 0){stop("The variable 'election'and 'unit' must not have NA values.", call. = FALSE)}
        if(length(method) > 1){stop("you must select only one method.", call. = FALSE)}

        ch.met <- c("Mainwaring and Jones", "Chibber and Kollman")
        nu.met <- c(1, 2)
        if(!any(method == ch.met | method == nu.met)){stop("the selected method does not exist.", call. = FALSE)}
        if(scale != 100 && scale != 1){stop("The value of 'scale' is not correct.", call. = FALSE)}
        vscale <- unlist(lapply(split(tidy_data, tidy_data$election), function(x){split(x, x$unit)}), recursive = FALSE)
        tidy_data <- lapply(vscale, function(x){cbind.data.frame(x,
                                                                 t.votes     = (x$votes/sum(x$votes, na.rm=TRUE))*scale,
                                                                 t.votes_nac = (x$votes_nac/sum(x$votes_nac, na.rm=TRUE))*scale )})
        tidy_data <- do.call(rbind, lapply(tidy_data, "[", -c(4:5)))
        rownames(tidy_data) <- NULL

        if(method == "Mainwaring and Jones" || method == 1){

                v1   <- unlist(lapply(split(tidy_data, tidy_data$election), function(x){split(x, x$party)}), recursive = FALSE)
                v2   <- lapply(v1, function(x){cbind(x, pns = apply(x[4], 2, gini_esaps))})
                pns1 <- do.call(rbind, lapply(v2, function(x){x[1,-2]}))
                psns <- lapply(split(pns1, pns1$election), function(x){cbind(x, psns = round(sum(x$t.votes_nac*x$pns),3))})
                psns <- do.call(rbind,lapply(psns, "[", 1, c(1, 6)))
                rownames(psns) <- NULL

                if(isTRUE(pns)){
                        pn <- pns1[order(pns1$party), c(2,1,5)]
                        rownames(pn) <- NULL
                        return(list(PSNS = psns, PNS = pn))
                }
                return(psns)
        }

        if(method == "Chibber and Kollman" || method == 2){
                nep_nac <- lapply(lapply(split(tidy_data, tidy_data$election),
                                         function(x){x[duplicated(x$party) == FALSE, ]}),
                                                function(x){ENP(x$t.votes_nac)})
                nep_loc <- unlist(lapply(split(tidy_data, tidy_data$election), function(x){split(x, x$unit)}), recursive = FALSE)
                nep_loc <- do.call(rbind,lapply(nep_loc, function(x){cbind(x, nepl = ENP(x$t.votes))}))
                out     <- lapply(split(nep_loc, nep_loc$election), function(x){mean(x[duplicated(x$unit) == FALSE, "nepl"])})
                output  <- data.frame(CH_K = round(unlist(nep_nac)-unlist(out),3))
                output

        }

}

































R/utils.R

calculus <- function(dat, digits){
        volatA <- dat
        for(i in 1:length(volatA)){
                for(j in 1:length(volatA[[i]])){
                        for(k in 1:(ncol(volatA[[i]][[j]])-2)){
                                volatA[[i]][[j]][2,(k+2)] <- abs(volatA[[i]][[j]][1,(k+2)] - volatA[[i]][[j]][2,(k+2)])
                        }
                        volatA[[i]][[j]]$volat_A <- round((1/2)*sum(volatA[[i]][[j]][2,3:ncol(volatA[[i]][[j]])]), digits)
                        volatA[[i]][[j]] <- volatA[[i]][[j]][2,c(1,2,ncol(volatA[[i]][[j]]))]
                }
                volatA[[i]]<-do.call(rbind, volatA[[i]])
        }
        volatA <- do.call(rbind, volatA)
        return(volatA)
}


#ENP  <- function(x){1/sum((x/sum(x))^2)}
ENP  <- function(x){1/sum((x/sum(x, na.rm = TRUE))^2, na.rm = TRUE)}

gini_esaps <- function(v){
        v <- ifelse(is.na(v), 0, v)
        v <- v[order(v)]
        l <- length(v)
        g <- sum(v*1L:l)
        g <- 2 * g/sum(v) - (l + 1L)
        round(1-(g/l), 3)
}


if(getRversion() >= "2.15.1"){utils::globalVariables(c('digits'))}







R/convert_esaps.R
#' @title Converts data in table form to tidy_data
#'
#' @description Convert data in table format to tidy_data to use in the indicators of
#'     the tidy_data.
#' @param path Character vector containing one or more path names.
#' @param dataset Electoral results by party. It can be a \code{data.frame} or a \code{list}.
#' @param file.name Name of the data file with extension.
#' @param nSheets Number of countries (number of sheets).'
#'     'Country' is a generic unit of analysis (province, department, etc.)
#' @param election.name Name of the variable that contains elections.
#' @param unit.name Name of the variable that contains the unit.
#'     'unit' is a generic unit of analysis (province, department, etc.)
#' @param M.name Name of the variable that contains the district magnitude
#'     (M+1). It is for the calculation of endogenous and exogenous
#'     electoral volatility (Torcal and Lago, 2015).
#' @param votes_nac.name la la la
#' @param allSheet By default it is \code{FALSE}. Load all the sheets that are in the
#'     files selected in \code{file.name}. This argument takes precedence over \code{nSheets}.
#' @param seats By default it is \code{FALSE}. If it is \code{TRUE}, it indicates that, in addition
#'     to electoral data per party, there is allocation data per seat. In this case,
#'     one column must be loaded for each party with the electoral result and another
#'     with the number of seats it obtained. The structure must be:
#'     party_1, party_2, ..., party_n, seats_party_1, seats_party_2, ..., seats_party_n.
#'
#'
#' @return data.frame
#'
#' @author Nicolas Schmidt \email{nschmidt@cienciassociales.edu.uy}
#'
#' @examples
#' votes <- list(data.frame(country = rep("ARG", 3),
#'                          year = c(1995, 2000, 2005),
#'                          party_A = c(40,10,20),
#'                          party_B = c(35,20,40),
#'                          party_C = c(25,70,40)),
#'               data.frame(country = rep("URY", 4),
#'                          year = c(1995, 2000, 2005, 2010),
#'                          party_A = c(30,30,20,20),
#'                          party_B = c(30,50,40, 30),
#'                          party_C = c(30,10,30, 25),
#'                          party_D = c(10,10,10,25)),
#'               data.frame(country = rep("BRA", 2),
#'                          year = c(1993, 1998),
#'                          party_A = c(30, 55),
#'                          party_B = c(70, 45)))
#'
#' votes <- convert_esaps(dataset = votes, unit.name = "country", election.name = "year")
#'
#'
#' votes2 <- data.frame(year = c(2000, 2005),
#'                      country = "URY",
#'                      votes_party1 = c(20, 30),
#'                      votes_party2 = c(30, 35),
#'                      votes_party3 = c(40, 25),
#'                      votes_party4 = c(10, 10),
#'                      seats_party1 = c(25, 35),
#'                      seats_party2 = c(20, 30),
#'                      seats_party3 = c(40, 30),
#'                      seats_party4 = c(15, 5))
#'
#' votes <- convert_esaps(dataset = votes2, unit.name = "country", election.name = "year",
#'                        seats = TRUE)
#'
#'
#' \dontrun{
#' v1 <- convert_esaps(path = getwd(),
#'                    file.name = c("electionBRA.xlsx", "electionARG.xlsx"),
#'                    election.name = "elec",
#'                    unit.name = "district",
#'                    allSheet = TRUE)
#'
#' v2 <-convert_esaps(path = getwd(),
#'                    file.name = c("ARG.ods", "URY.ods", "BRA.ods"),
#'                    nCountry = c(2, 3, 1),
#'                    election.name = "elec",
#'                    unit.name = "province")
#'
#' v3 <- convert_esaps(path = here::here(),
#'                    file.name = list.files(pattern = "*.xlsx"),
#'                    election.name = "year",
#'                    unit.name = "country",
#'                    M.name = "magnitude",
#'                    seats = TRUE,
#'                    allSheet = TRUE)
#' }
#'
#' @export


convert_esaps<-function(path = NULL,
                        dataset = NULL,
                        file.name = NULL,
                        nSheets = 1,
                        election.name,
                        unit.name,
                        M.name = NULL,
                        votes_nac.name = NULL,
                        seats = FALSE,
                        allSheet = FALSE){

        if (is.null(path) && is.null(dataset)){
                stop("any of the arguments 'path' or 'dataset' must not be NULL.", call. = FALSE)
        }
        if (!is.null(path)) {
                extention <- c("xlsx", "ods")
                extention <- extention[c(all(grepl("xlsx$", file.name)==TRUE), all(grepl("ods$", file.name)==TRUE))]
                if (length(extention) != 1L){
                        stop("It only accepts extension 'xlsx' or 'ods'.", call. = FALSE)
                }
        if(length(file.name) != length(nSheets)){
                nSheets <- rep(nSheets[1], length(file.name))
        }
        old_setwd <- getwd()
        setwd(path)
        if (extention == "ods") {
                if(isTRUE(allSheet)){
                        nSheets <- vector("numeric", length = length(file.name))
                        for(i in 1:length(file.name)){
                                nSheets[i] <- length(readODS::ods_sheets(file.name[i]))
                        }
                }
                dat <- list()
                init <- 0L
                for(i in 1:length(file.name)){
                        for (j in 1:nSheets[i]) {
                                dat[[j+init]] <- as.data.frame(readODS::read_ods(file.name[i], sheet = j))
                        }
                        init <- length(dat)
                }
        }
        if (extention == "xlsx") {
                if(isTRUE(allSheet)){
                        nSheets <- vector("numeric", length = length(file.name))
                        for(i in 1:length(file.name)){
                                nSheets[i] <- length(readxl::excel_sheets(file.name[i]))
                                }
                        }
                        dat <- list()
                        init <- 0L
                        for(i in 1:length(file.name)){
                                for (j in 1:nSheets[i]) {
                                        dat[[j+init]] <- as.data.frame(readxl::read_excel(file.name[i], sheet = j))
                                }
                                init <- length(dat)
                        }
                }
                setwd(old_setwd)
        }
        if (!is.null(dataset)) {
                if (is.matrix(dataset)) {stop("'dataset' must be a list or a data.frame.", call. = FALSE)}
                if (is.list(dataset) && !is.data.frame(dataset)) {dat <- dataset}
                if (is.data.frame(dataset)) {dat <- list(dataset)}
        }
        if (0L %in% sapply(dat, nrow)) {
                zero <- sapply(dat, nrow) %in% 0L
                if(length(zero)==sum(zero)){
                        stop("The datasets have 0 rows.", call. = FALSE)
                }else{
                        dat <- dat[!zero]
                }
        }
        if(sum(sapply(dat, function(x) c(unit.name, election.name) %in% names(x))) == 0){
                stop("'election.name' and 'unit.name' must be the same on all sheets.", call. = FALSE)
        }
        vector1 <- numeric()
        vector2 <- numeric()
        for (i in 1:length(dat)) {
                vector1[i] <- which(colnames(dat[[i]]) == election.name)
                colnames(dat[[i]])[vector1[i]] <- "election"
                vector2[i] <- which(colnames(dat[[i]]) == unit.name)
                colnames(dat[[i]])[vector2[i]] <- "unit"
        }
        dat2 <- list()
        dat3 <- list()
        out  <- list()
        for (i in 1:length(dat)) {
                dat2[[i]] <- dat[[i]][ c(vector1[i], vector2[i])]
                dat3[[i]] <- dat[[i]][-c(vector1[i], vector2[i])]
                out[[i]] <- cbind(dat2[[i]], dat3[[i]])
                out[[i]]$unit <- as.character(out[[i]]$unit)
        }
        if (!is.null(M.name)) {

                if(!all(sapply(out, function(x) M.name %in% names(x)))){stop("'M.name' must be the same on all sheets.", call. = FALSE)}
                m <- numeric()
                data2 <- list()
                for (i in 1:length(out)) {
                        m[i] <- which(names(out[[i]]) == M.name)
                        data2[[i]] <- out[[i]][,m[i]]
                        out[[i]] <- out[[i]][,-m[i]]
                        out[[i]] <- cbind(out[[i]][1:2], M = data2[[i]], out[[i]][3:ncol(out[[i]])])
                }
        }
        if (!is.null(votes_nac.name)) {
                if(isTRUE(seats) || !is.null(M.name)){stop("if 'votes_nac.name' are different from NULL, 'M.name' or 'seats' must have the default values.", call. = FALSE)}
                out0<- lapply(out, function(x){tidyr::gather(x[which(x[,2]==votes_nac.name),-2], "party", "votes_nac", -"election")})
                out <- lapply(out, function(x){tidyr::gather(x[-which(x[,2]==votes_nac.name),], "party", "votes", -"election", -"unit")})
                output <- list()
                for(i in 1:length(out)){output[[i]] <- dplyr::inner_join(out[[i]],out0[[i]] , by=c("election", "party"))}
                pull <- do.call(rbind, output)
                rownames(pull) <- NULL
                return(pull)
        }
        variables <- c(election.name, unit.name, M.name) ## aca no va votes_nac_name???
        var <- length(variables)
        if(var > 2){variables <- c("election", "unit", "M")}else{variables <- c("election", "unit")}

        if(isTRUE(seats)){
                larS <- length(out)
                c_seats <- as.logical((sapply(out, ncol)-var) %% 2)
                if(length(which(c_seats==TRUE)) > 0){out <- out[-c(which(c_seats==TRUE))]}
                if(length(out)!=larS && length(out) != 0){warning("The database was deleted: ", paste(which(c_seats==TRUE), collapse = ", "), " ...must have the same number of columns of parties and seats", call. = FALSE)}
                if(length(out)==0L){stop("The structure of the data is not correct.", call. = FALSE)}
                out3 <- lapply(out, function(x){x[,c(1:2,(((ncol(x)-var)/2)+(var+1)):ncol(x))]})
                out4 <- lapply(out, function(x){x[,-c((((ncol(x)-var)/2) + (var+1)): ncol(x))]})
                out3 <- lapply(out3, function(x){tidyr::gather(x, "party", "seats", -"election", -"unit")})
                out4 <- lapply(out4, function(x){tidyr::gather(x, "party", "votes", -variables)})
                for(i in 1:length(out4)){out4[[i]]<-cbind(out4[[i]], seats = out3[[i]]$seats)}
                pull <- do.call(rbind, out4)
        }else{
                out4 <- lapply(out, function(x){tidyr::gather(x, "party", "votes", -variables)})
                pull <- do.call(rbind.data.frame, out4)
        }
        if(length(unique(as.character(pull$unit)))>1){
                pull <- pull[order(pull$unit), ]
        }else{
                pull <- pull[order(pull$election), ]
        }
        rownames(pull) <- NULL
        return(pull)
}






